feat: add accessor protocol and refactor `stats/base/nanmskmax`

rahulptl165 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: na
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves #5662 .

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds new files `accessors.js`.
-   Modified existing `nanmskmax.js`.
-   Modified existing `ndarray.js`.
-   Modified readme.md, benchmark, docs/types, docs/repl.txt, examples, tests files.

### Important
```
function nanmskmax( N, x, strideX, mask, strideMask ) {
	return ndarray( N, x, strideX, stride2offset( N, strideX ), mask, strideMask, stride2offset( N, strideMask ) );
}
```
Instead of above code in `nanmskmax.js` and tests files, I have used below code due to eslint warnings about maximum character allowed in a single line.
```
function nanmskmax( N, x, strideX, mask, strideMask ) {
	var offsetMask = stride2offset( N, strideMask );
	var offsetX = stride2offset( N, strideX );
	return ndarray( N, x, strideX, offsetX, mask, strideMask, offsetMask );
}
```
If we should not bother about warnings, request changes in review, i will modify it.
## Related Issues

This pull request:

-   resolves #5662 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

rahulptl165 committed:
feat: add accessor protocol and refactor stats/base/nanmskmax

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmskmax/README.md:
```diff
@@ -44,28 +44,25 @@ Computes the maximum value of a strided array `x` according to a `mask`, ignorin
+-   **strideMask**: stride length for `mask`.
 
-The `N` and `stride` parameters determine which elements are accessed at runtime. For example, to compute the maximum value of every other element in `x`,
+The `N` and stride parameters determine which elements int the strided arrays are accessed at runtime. For example, to compute the maximum value of every other element in `x`,
```
```suggestion
The `N` and stride parameters determine which elements in the strided arrays are accessed at runtime. For example, to compute the maximum value of every other element in `x`,
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmskmax/README.md:
````diff
@@ -44,28 +44,25 @@ Computes the maximum value of a strided array `x` according to a `mask`, ignorin
 ```javascript
-var floor = require( '@stdlib/math/base/special/floor' );
-
 var x = [ 1.0, 2.0, -7.0, -2.0, 4.0, 3.0, 5.0, 6.0 ];
````
```suggestion
var x = [ 1.0, 2.0, -7.0, -2.0, 4.0, 3.0, 5.0, 6.0, NaN ];
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmskmax/README.md:
```diff
@@ -44,28 +44,25 @@ Computes the maximum value of a strided array `x` according to a `mask`, ignorin
-var floor = require( '@stdlib/math/base/special/floor' );
-
 var x = [ 1.0, 2.0, -7.0, -2.0, 4.0, 3.0, 5.0, 6.0 ];
 var mask = [ 0, 0, 0, 0, 0, 0, 1, 1 ];
```
```suggestion
var mask = [ 0, 0, 0, 0, 0, 0, 1, 1, 1 ];
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmskmax/README.md:
```diff
@@ -44,28 +44,25 @@ Computes the maximum value of a strided array `x` according to a `mask`, ignorin
-var N = floor( x.length / 2 );
 
-var v = nanmskmax( N, x, 2, mask, 2 );
+var v = nanmskmax( 4, x, 2, mask, 2 );
```
```suggestion
var v = nanmskmax( 5, x, 2, mask, 2 );
```
We should explicitly traverse over the `NaN` element to show how it is being ignored by the algorithm

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmskmax/README.md:
```diff
@@ -76,17 +73,14 @@ Note that indexing is relative to the first index. To introduce offsets, use [`t
 var Uint8Array = require( '@stdlib/array/uint8' );
-var floor = require( '@stdlib/math/base/special/floor' );
 
 var x0 = new Float64Array( [ 2.0, 1.0, -2.0, -2.0, 3.0, 4.0, 5.0, 6.0 ] );
```
same comment regarding `NaN` values in examples

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmskmax/docs/repl.txt:
```diff
@@ -104,14 +103,13 @@
     2.0
 
     // Using offset parameter:
     > x = [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0, 4.0 ];
```
same comment regarding `NaN` values in exaamples

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmskmax/docs/types/test.ts:
```diff
@@ -26,7 +27,7 @@ import nanmskmax = require( './index' );
 	const mask = new Uint8Array( 10 );
 
-	nanmskmax( x.length, x, 1, mask, 1 ); // $ExpectType number
+	nanmskmax( x.length, new AccessorArray( x ), 1, mask, 1 ); // $ExpectType number
```
```suggestion
	nanmskmax( x.length, x, 1, mask, 1 );
	nanmskmax( x.length, new AccessorArray( x ), 1, mask, 1 ); // $ExpectType number
```
don't remove the old test, add a new one for accessor arrays

aayush0325 replied:
same comment for the `ndarray` API below

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmskmax/test/test.ndarray.js:
```diff
@@ -164,16 +221,96 @@ tape( 'the function supports `stride` parameters', function test( t ) {
+
+	x = toAccessorArray( x );
+
+	v = nanmskmax( 5, x, 2, 0, toAccessorArray( mask ), 2, 0 );
```
```suggestion
	v = nanmskmax( 6, x, 2, 0, toAccessorArray( mask ), 2, 0 );
```
follow the commented indices given above

aayush0325 replied:
same comment for all other instances

aayush0325 commented:
kindly also adhere to the file structure that we're trying to follow, your `lib` directory should only contain a `main.js`, `ndarray.js`, `accessors.js` and `index.js`, see #6248 for reference

rahulptl165 committed:
refactor: refactored lib folder's file to align with existing style

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

rahulptl165 commented:
Hi @aayush0325, I have addressed requested changes. pls review it.

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmskmax/test/test.ndarray.js:
```diff
@@ -206,16 +343,56 @@ tape( 'the function supports negative `stride` parameters', function test( t ) {
 	t.end();
 });
 
 tape( 'the function supports `offset` parameters', function test( t ) {
```
```suggestion
tape( 'the function supports `offset` parameters (accessors)', function test( t ) {
```

kindly fix all other errors similar to this in the tests, it makes it hard for someone else to go through these if the descriptions are not accurate

rahulptl165 committed:
docs: update test description for better clearity

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
Merge branch 'develop' into feat/accessors-nanmskmax

Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

gururaj1512 commented:
/stdlib merge

gururaj1512 committed:
docs: update README for clarity and formatting

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
docs: update benchmark files for consistency and clarity

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: update example usage

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
docs: update documentation for nanmskmax function

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
test: fix accessor test cases

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
docs: update examples in nanmskmax

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: examples

Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

gururaj1512 committed:
fix: examples in README

Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

gururaj1512 committed:
refactor: benchmark

Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

gururaj1512 committed:
refactor: benchmark

Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmskmax/README.md:
````diff
@@ -52,41 +52,35 @@ The function has the following parameters:
 ```javascript
-var floor = require( '@stdlib/math/base/special/floor' );
+var x = [ 1.0, 2.0, -7.0, -2.0, 4.0, 3.0, 5.0, 6.0, NaN, NaN ];
+var mask = [ 0, 0, 0, 0, 0, 0, 1, 1, 1, 1 ];
````
```suggestion
var mask = [ 0, 0, 0, 0, 0, 0, 1, 1, 0, 0 ];
```

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmskmax/README.md:
```diff
@@ -52,41 +52,35 @@ The function has the following parameters:
 var x1 = new Float64Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
 
-var mask0 = new Uint8Array( [ 0, 0, 0, 0, 0, 0, 1, 1 ] );
+var mask0 = new Uint8Array( [ 0, 0, 0, 0, 0, 0, 1, 1, 1, 1 ] );
```
```suggestion
var mask0 = new Uint8Array( [ 0, 0, 0, 0, 0, 0, 1, 1, 0, 0 ] );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmskmax/README.md:
```diff
@@ -107,16 +101,13 @@ The function has the following additional parameters:
 -   **offsetMask**: starting index for `mask`.
 
-While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the maximum value for every other value in `x` starting from the second value
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter support indexing semantics based on a starting indices. For example, to calculate the maximum value for every other element in `x` starting from the second element
```
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to calculate the maximum value for every other element in `x` starting from the second element
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmskmax/docs/repl.txt:
```diff
@@ -108,10 +107,9 @@
-    > var N = {{alias:@stdlib/math/base/special/floor}}( x.length / 2 );
-    > {{alias}}.ndarray( N, x, 2, 1, mask, 2, 1 )
+    > x = [ 1.0, -2.0, 3.0, 2.0, 5.0, -1.0, 4.0, NaN, NaN ];
+    > mask = [ 0, 0, 0, 0, 0, 0, 1, 1, 1 ];
```
```suggestion
    > mask = [ 0, 0, 0, 0, 0, 0, 1, 0, 0 ];
```

kgryte committed:
docs: update example

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmskmax/lib/accessors.js:
```diff
@@ -0,0 +1,109 @@
+/**
+* Computes the maximum value of a strided array according to a mask, ignoring `NaN` values.
+*
+* @param {PositiveInteger} N - number of indexed elements
```
```suggestion
* @private
* @param {PositiveInteger} N - number of indexed elements
```

kgryte committed:
docs: fix missing private annotation

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmskmax/test/test.main.js:
```diff
@@ -117,6 +175,23 @@ tape( 'if provided an `N` parameter less than or equal to `0`, the function retu
+	var v;
+
+	x = toAccessorArray( [ 1.0, -2.0, -4.0, 5.0, 3.0 ] );
+	mask = toAccessorArray( [ 0.0, 0.0 ] );
```
```suggestion
	mask = toAccessorArray( [ 0, 0 ] );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmskmax/test/test.main.js:
```diff
@@ -131,9 +206,22 @@ tape( 'if provided an `N` parameter equal to `1`, the function returns the first
+	var v;
+
+	x = toAccessorArray( [ 1.0, -2.0, -4.0, 5.0, 3.0 ] );
+	mask = toAccessorArray( [ 0.0, 0.0 ] );
```
```suggestion
	mask = toAccessorArray( [ 0, 0 ] );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmskmax/test/test.ndarray.js:
```diff
@@ -115,6 +173,23 @@ tape( 'if provided an `N` parameter less than or equal to `0`, the function retu
+	var v;
+
+	x = toAccessorArray( [ 1.0, -2.0, -4.0, 5.0, 3.0 ] );
+	mask = toAccessorArray( [ 0.0, 0.0 ] );
```
```suggestion
	mask = toAccessorArray( [ 0, 0 ] );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmskmax/test/test.ndarray.js:
```diff
@@ -129,9 +204,22 @@ tape( 'if provided an `N` parameter equal to `1`, the function returns the first
+	var v;
+
+	x = toAccessorArray( [ 1.0, -2.0, -4.0, 5.0, 3.0 ] );
+	mask = toAccessorArray( [ 0.0, 0.0 ] );
```
```suggestion
	mask = toAccessorArray( [ 0, 0 ] );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
